src/xdp/program.c

```diff
@@ -1776,14 +1913,23 @@ XdpProgramAttach(
             }
         } else {
             //
-            // Revert changes to the program, scrap the metaprogram, and bail.
-            // This can happen only if no metaprogram was previously attached.
+            // Revert changes to the program, revert changes to the current
+            // metaprogram (if it exists), scrap the new metaprogram, and bail.
             //
-            ASSERT(ExistingProgramObject == NULL);
 
             RemoveEntryList(&ProgramObject->SharingLink);
             InitializeListHead(&ProgramObject->SharingLink);
 
+            if (ExistingProgramObject != NULL) {
+                ASSERT(ExistingProgramObject->Flags.IsMetaProgram);
+                ASSERT(IsListEmpty(&ExistingProgramObject->SharingLink));
+                ASSERT(!IsListEmpty(&NewMetaProgramObject->SharingLink));
+                AppendTailList(
+                    &ExistingProgramObject->SharingLink, &NewMetaProgramObject->SharingLink);
+                RemoveEntryList(&NewMetaProgramObject->SharingLink);
+                InitializeListHead(&NewMetaProgramObject->SharingLink);
+            }
```

Comment:
Is this a bugfix?

Reply:
No, this is added because of a contract change that allows `XdpRxQueueSetProgram` above to fail when a program is already attached. Previously, if `ExistingProgramObject` were NULL (i.e. a program is already attached) then `XdpRxQueueSetProgram` was required to succeed.

I made this contract change because we need to reject the new program if it uses the new TX action and the underlying interface queue it is attaching to does not support the TX action. Ideally we'd validate that ahead of time, but it turns out that needs a more significant refactoring of the RX queue <-> interface queue binding model.

Reply:
> The spinxsk test is still hitting hangs in local testing, but otherwise this PR should be ready for review.

I fixed one bug (potentially returning an empty list to NDIS) that seems to have fixed spinxsk, though I'm honestly not quite sure how. I've kicked off an internal long-haul spinxsk test for extra coverage.